### ISSUE TITLE
Pylint 2.7, astroid 2.5 unsubscriptable object for Tuple coming from a modifier

### ISSUE THREAD

**Issue by LefterisJP** [PyCQA/pylint#4114]
### Steps to reproduce
Upgraded pylint for our project to latest release

```diff
-pylint==2.6.2
-astroid==2.4.2  # engine of pylint, upgrade them together
+pylint==2.7.0
+astroid==2.5  # engine of pylint, upgrade them together
```

Got some new errors which I believe to be false positives. They all stem from calls into the same function of a library we use. I made a simplified test file to demonstrate. I think it's all due the modifier.

```python
import functools
from typing import Callable, TypeVar

T = TypeVar("T")


def apply_to_return_value(
    callback: Callable[..., T]
) -> Callable[..., Callable[..., T]]:
    def outer(fn: Callable[..., T]) -> Callable[..., T]:
        @functools.wraps(fn)
        def inner(*args, **kwargs) -> T:  # type: ignore
            return callback(fn(*args, **kwargs))

        return inner

    return outer

to_tuple = apply_to_return_value(tuple)

class Decoder():

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __call__(self):
        return Decoder.decode([self.x, self.y])

    @to_tuple
    def decode(data):
        return data



def foo():
    decoder = Decoder(5, 6)
    return decoder()


a = foo()
print(a[1])  # <--  Value 'a' is unsubscriptable (unsubscriptable-object)
```

### Current behavior

Pylint sees `a` as an unsubscriptable object

### Expected behavior

That should not happen. It seems to only happen due to the modifier not being understood by the pylint parser.

### pylint --version output

```
pylint 2.7.0
astroid 2.5
Python 3.7.9 (default, Sep 26 2020, 15:11:06) 
[GCC 10.2.0]
```

It's perhaps noteworthy to say that this did not happen in our project with pylint 2.6.2 and astroid 2.4.2

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue, I can reproduce the problem from your example.

**Comment by cdce8p** [comment]
This issue seems to be fixed when testing with the latest pylint version.
If you still encounter issues, please report back and I'll reopen it.
